File: app/src/main/java/com/pluu/support/olleh/OllehEpisodeApi.java
package com.pluu.support.olleh;

import android.content.Context;

import com.pluu.support.impl.AbstractEpisodeApi;
import com.pluu.webtoon.item.Episode;
import com.pluu.webtoon.item.EpisodePage;
import com.pluu.webtoon.item.WebToonInfo;
import com.pluu.webtoon.item.WebToonType;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * 올레 웹툰 Episode API
 * Created by PLUUSYSTEM-NEW on 2015-10-31.
 */
public class OllehEpisodeApi extends AbstractEpisodeApi {

	private static final String EPISODE_URL = "http://webtoon.olleh.com/api/work/getTimesListByWork.kt";
	private final int PAGE_SIZE = 20;

	private JSONArray savedArray;
	private int totalSize;

	private Episode firstEpisode;
	private int page = 0;

	private String id;

	public OllehEpisodeApi(Context context) {
		super(context);
	}

	@Override
	public EpisodePage parseEpisode(WebToonInfo info) {
		id = info.getToonId();

		EpisodePage episodePage = new EpisodePage(this);

		try {
			String response;
			if (savedArray == null) {
				response = requestApi();
				savedArray = new JSONObject(response).optJSONArray("timesList");
				totalSize = savedArray.length();
				if (savedArray != null) {
					totalSize = savedArray.length();
					firstEpisode
						= createEpisode(info, savedArray.optJSONObject(totalSize - 1));
				}
			}

			if (totalSize > 0) {
				episodePage.episodes = parseList(info, savedArray, page);
				episodePage.nextLink = getNextPageLink(totalSize, page);
			}
			page++;
		} catch (Exception e) {
			e.printStackTrace();
		}
		return episodePage;
	}

	private List<Episode> parseList(WebToonInfo info, JSONArray array, int page) {
		List<Episode> list = new ArrayList<>();

		int startPage = page * PAGE_SIZE;
		int endPage = (page + 1) * PAGE_SIZE;
		if (endPage > totalSize) {
			endPage = totalSize;
		}

		for (int i = startPage; i < endPage; i++) {
			list.add(createEpisode(info, array.optJSONObject(i)));
		}
		return list;
	}

	private Episode createEpisode(WebToonInfo info, JSONObject obj) {
		Episode item = new Episode(info, obj.optString("timesseq"));
		item.setEpisodeTitle(obj.optString("timestitle"));
		item.setImage(info.getType() == WebToonType.TOON
						  ? obj.optString("thumbpath")
						  : info.getImage());
		item.setRate(obj.optString("totalstickercnt"));
//				item.setUpdateDate(obj.optString("regdt"));

		return item;
	}

	private String getNextPageLink(int totalSize, int current) {
		int total = (int) Math.ceil(totalSize / (double) PAGE_SIZE);
		if (total >= current + 1) {
			// 다음 페이지 존재
			return EPISODE_URL;
		}

		return null;
	}

	@Override
	public String moreParseEpisode(EpisodePage item) {
		return item.nextLink;
	}

	@Override
	public Episode getFirstEpisode(Episode item) {
		return firstEpisode;
	}

	@Override
	public void init() {
		super.init();
		savedArray = null;
		page = 0;
	}

	@Override
	public String getMethod() {
		return POST;
	}

	@Override
	public String getUrl() {
		return EPISODE_URL;
	}

	@Override
	public Map<String, String> getHeaders() {
		Map<String, String> map = new HashMap<>();
		map.put("Referer", "http://webtoon.olleh.com");
		return map;
	}

	@Override
	public Map<String, String> getParams() {
		Map<String, String> map = new HashMap<>();
		map.put("mobileyn", "N");
		map.put("toonfg", "toon");
		map.put("sort", "subject");
		map.put("webtoonseq", id);
		return map;
	}

}

File: app/src/main/java/com/pluu/support/olleh/OllehEpisodeApi.kt
package com.pluu.support.olleh

import android.content.Context
import com.pluu.support.impl.AbstractEpisodeApi
import com.pluu.support.impl.NetworkSupportApi
import com.pluu.webtoon.item.Episode
import com.pluu.webtoon.item.EpisodePage
import com.pluu.webtoon.item.WebToonInfo
import com.pluu.webtoon.item.WebToonType
import org.json.JSONArray
import org.json.JSONObject

/**
 * 올레 웹툰 Episode API
 * Created by pluu on 2017-04-22.
 */
class OllehEpisodeApi(context: Context) : AbstractEpisodeApi(context) {
    private val PAGE_SIZE = 20

    private var savedArray: JSONArray? = null
    private var totalSize: Int = 0

    private var firstEpisode: Episode? = null
    private var page = 0

    private lateinit var id: String

    override fun parseEpisode(info: WebToonInfo): EpisodePage {
        id = info.toonId

        if (savedArray == null) {
            try {
                bindRequest(info)
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }

        val episodePage = EpisodePage(this).apply {
            episodes = parseList(info, savedArray, page)
            nextLink = getNextPageLink(totalSize, page)
        }
        page++
        return episodePage
    }

    @Throws(Exception::class)
    private fun bindRequest(info: WebToonInfo) {
        val response = requestApi()
        savedArray = JSONObject(response).optJSONArray("timesList")?.apply {
            totalSize = length()
            firstEpisode = createEpisode(info, optJSONObject(totalSize - 1))
        }
    }

    private fun parseList(info: WebToonInfo, array: JSONArray?, page: Int): List<Episode> {
        if (array == null) {
            return emptyList()
        }

        val list = mutableListOf<Episode>()
        val startPage = page * PAGE_SIZE
        var endPage = (page + 1) * PAGE_SIZE
        if (endPage > totalSize) {
            endPage = totalSize
        }

        (startPage until endPage)
                .mapTo(list) { createEpisode(info, array.optJSONObject(it)) }
        return list
    }

    private fun createEpisode(info: WebToonInfo, obj: JSONObject): Episode {
        return Episode(info, obj.optString("timesseq")).apply {
            episodeTitle = obj.optString("timestitle")
            image = if (info.type == WebToonType.TOON)
                obj.optString("thumbpath")
            else
                info.image
            rate = obj.optString("totalstickercnt")
        }
    }

    private fun getNextPageLink(totalSize: Int, current: Int): String? {
        val total = Math.ceil(totalSize / PAGE_SIZE.toDouble()).toInt()
        if (total >= current + 1) {
            // 다음 페이지 존재
            return EPISODE_URL
        }
        return null
    }

    override fun moreParseEpisode(item: EpisodePage): String {
        return item.nextLink
    }

    override fun getFirstEpisode(item: Episode): Episode? {
        return firstEpisode
    }

    override fun init() {
        super.init()
        savedArray = null
        page = 0
    }

    override val method: String
        get() = NetworkSupportApi.POST

    override val url: String
        get() = EPISODE_URL

    override val headers: Map<String, String>
        get() = hashMapOf("Referer" to OllehWeekApi.HOST)

    override val params: Map<String, String>
        get() = hashMapOf("webtoonseq" to id)

    companion object {
        private val EPISODE_URL = OllehWeekApi.HOST + "/api/work/getTimesListByWork.kt"
    }

}
